### PR TITLE
Add agent enum with string function and convert abs to map

### LINKED ISSUE #91: agentType is wrong in `NewBaseAgent`
`agentType` does not correspond to the team number currently - I'm not sure what the intended behaviour here is. Currently it's an `int` and I don't think it should be an int.

### PR DESCRIPTION
**Pull request by jzzheng22**
# Summary
Closes #91.

Adds a new agent type for type safety, with an enum.
The list of agent constructors is now a map inside simulation environment.
Changes the lists for agent counts to a map where the key is an agent type.
Tidies up `generateInitialAgents`
Update logger functions to use stringify function
Update `utilFunctions.Sum()` to take map instead of list.

## Test Plan
```
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team1Agent1"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team1Agent1"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team4"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team4"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"RandomAgent"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"RandomAgent"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team1Agent2"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team1Agent2"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team2"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team2"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team3"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team3"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team5"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team5"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team6"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team6"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team7"}
{"level":"info","msg":"Creating new agent","reporter":"simulation","time":"2021-12-31T18:42:51Z","type":"Team7"}
```

```
{"agent_id":"f37bafb5-35da-424a-84da-4269e4bd35e7","agent_type":"Team1Agent2","level":"info","msg":"Custom agent reporting status without using fields","reporter":"agent","time":"2021-12-31T18:42:51Z"}
{"agent_id":"f37bafb5-35da-424a-84da-4269e4bd35e7","agent_type":"Team1Agent2","level":"info","msg":"I got no thing","reporter":"agent","time":"2021-12-31T18:42:51Z"}
{"agent_id":"61c765ea-b38a-4c8c-94d7-f08023e98d14","agent_type":"Team7","aggr":79,"floor":14,"greed":21,"hp":100,"kind":79,"level":"info","msg":"Team7Agent1 reporting status:","reporter":"agent","time":"2021-12-31T18:42:51Z"}
{"agent_id":"225792b4-8860-4dcb-b236-a9d2a91d225b","agent_type":"Team1Agent2","level":"info","msg":"Custom agent reporting status without using fields","reporter":"agent","time":"2021-12-31T18:42:51Z"}
```

### CODE REVIEW

**Review by udai-arneja — APPROVED** [review]
LGTM, I think we could also move some of the agent stuff into the agent global type for nice structure. But can be done in a different commit for speed.

File: pkg/simulation/simulation.go [hunk 1]
```diff
@@ -36,6 +48,17 @@ func NewSimEnv(parameters *config.ConfigParameters, healthInfo *health.HealthInf
 		AgentsPerFloor: parameters.AgentsPerFloor,
 		logger:         *log.WithFields(log.Fields{"reporter": "simulation"}),
 		stateLog:       logging.NewLogState(parameters.LogFileName),
+		agentNewFuncs: map[agent.AgentType]AgentNewFunc{
+			agent.Team1Agent1: agent1.New,
```
Comment (VFjr) [inline pkg/simulation/simulation.go:52]: This is not really your problem, but can we rename team1's agent1 and agent2 packages to team1Agent1 and team1Agent2 so it is consistent with the naming of all other agents ? Or should I just create an issue for them so we don't touch code we don't own and tell them to do it themselves?

Comment (VFjr) [reply pkg/simulation/simulation.go:52]: Ah, I was too slow, I'll just make an issue

Comment (udai-arneja) [reply pkg/simulation/simulation.go:52]: I think things like this should be done in different commits, just to get things moving faster.

Comment (VFjr) [reply pkg/simulation/simulation.go:52]: Also just realised that these are the test agents, so it is not really a problem to have a different naming convention. Also I didn't even realise I am in Team1 lol.

**Pull request merged**